### ISSUE TITLE
Noop build performance regression after 8.6-rc-1 upgrade

### ISSUE THREAD

**Issue by bingranl** [gradle/gradle#27917]
### Current Behavior

We received build performance regression alerts for the TOTAL_BUILD_TIME of UP-TO-DATE build from multiple benchmark projects (https://issuetracker.google.com/issues/321137534)  

I add the step to reproduce using Gradle Profiler.  However, with Gradle profiler, the difference between 8.4 & 8.6-rc-1 is less obvious then we saw on our end. (see the numbers on our ticket)

### Expected Behavior

No build time increase 

### Steps to Reproduce

1. Clone https://github.com/bingranl/android-studio-gradle-test.3600041f which is obtained basically by extracting android-studio-gradle-test.3600041f and apply setup.diff
2. Benchmark with the scenario file, e.g. `ANDROID_HOME=/xxxx/sdk/linux LOCAL_REPO=/xxxx/android-studio-gradle-test.3600041f/repo gradle-profiler --benchmark --gradle-version 8.4 --gradle-version 8.6-rc-1 --scenario-file performance.scenarios upToDateSingleVariant`
3. See the result in profile-out-x folder, you should be able to see a 5% regression 

### Gradle version

8.4 and 8.6-rc-1

**Comment by bingranl** [comment]
Note: the max workers is set to 3

**Comment by ov7a** [comment]
Thank you for reporting this!

We'll look into reproducing these numbers on our side.

**Comment by ov7a** [comment]
@bingranl I tried to reproduce the numbers, but the variance of the output is too high to get a conclusive result.
It seems like the max workers setting does not affect the outcome.

[benchmark (1).csv](https://github.com/gradle/gradle/files/14178149/benchmark.1.csv)

Do you have any ideas where the variance might come from? Did you try to shrink the reproducer project to narrow down the root cause?

**Comment by bingranl** [comment]
If you compare the results from three builds for each row, it has some variance. However, if you compute the mean and median of all the builds of each Gradle version, the mean and median is both lower for the 8.4. 

I didn't try to shrink because I don't have much idea about where the problem could come from.

**Comment by ov7a** [comment]
If you look at min/max of run times:
<img width="602" alt="image" src="https://github.com/gradle/gradle/assets/4946730/4f4b09ae-c780-4d28-8ee7-d4404325fd25">
the min and max values can differ from the average up to 20%, so I don't think those numbers can be trusted.

**Comment by bingranl** [comment]
Yeah, the min and max differs a lot which is something we don't know why.

**Comment by ov7a** [comment]
This issue needs a decision from the team responsible for that area. They have been informed. Response time may vary.

***

@gradle/bt-configuration can you take a look at this during your evaluation of 8.7 performance?

**Comment by 6hundreds** [comment]
@bingranl I'm trying to profile the reproducer project, but not successfully, whereas benchmarking works. Have you tried? 
Also, is it possible to reduce amount of subprojects? You mentioned applying `setup.diff`, but I didn't find anywhere such a script.

**Comment by bingranl** [comment]
@6hundreds You can ignore setup.diff because it is a file used internally by us. If we reduce the amount of subprojects, it would be harder for you to see the performance difference. Can you share what error you are facing when running the benchmark project?

**Comment by 6hundreds** [comment]
@bingranl Got it. The benchmark is working. Profiling is doesn't, it fails with something like `Caused by: org.gradle.launcher.daemon.client.DaemonDisappearedException: Gradle build daemon disappeared unexpectedly (it may have been killed or may have crashed)`. It's probably an OOM, didn't investigated it deeply so far. Increasing xmx for the daemon is doesn't help tho.

**Comment by bamboo** [comment]
Is this still relevant, @bingranl?
